src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [Object.assign({}, state)];
+
+  for (const action of actions) {
+    const stateCopy = Object.assign({}, stateHistory[stateHistory.length - 1]);
+
+    if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        stateCopy[key] = action.extraData[key];
+      }
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete stateCopy[key];
+      }
+    } else if (action.type === 'clear') {
+      for (const key in stateCopy) {
+        delete stateCopy[key];
+      }
+    }
```

Comment:
checklist point 2: Use switch statement if you have limited amount of conditions